### PR TITLE
Relabel rewiring comments in COMPRESS-QUBITS

### PR DESCRIPTION
**Pull request by appleby**
The `COMPRESS-QUBITS` transform wasn't updating rewiring comments, which causes `PARSED-PROGRAM-TO-LOGICAL-MATRIX` to generate the wrong matrix if `COMPRESS-QUBITS` wasn't a no-op.

This was noticed while working on the fix for #361, which changes the default initial rewiring type for some programs from `':NAIVE` to `':PARTIAL` or `':GREEDY`. Making that change caused the tests in `QUILC-TESTS::TEST-FAITHFULNESS` to fail due to this `COMPRESS-QUBITS` bug.

~~This is still a WIP. Posting for early feedback to see if the basic approach seems sane.~~

Summary of changes in this PR:

1. Call RELABEL-REWIRING on rewiring comments in COMPRESS-QUBITS (see above)
2. Sprankle some abstraction over rewiring comments
3. Tests

TODOs:

- [x] Replace `READ-FROM-STRING` in `PARSE-ENTERING/EXITING-REWIRING` with something akin to `MAKE-REWIRING-FROM-STRING`
- [x] Refactor existing rewiring-related code to use new abstractions
- [x] Replace `COMPILER-HOOK` tests with more focused `COMPRESS-QUBITS` tests

### CODE REVIEW

File: app/src/entry-point.lisp [hunk 1]
```diff
@@ -490,7 +490,7 @@ Version ~A is available from https://www.rigetti.com/forest~%"
          (let* ((program-text (slurp-lines))
                 (program (parse-quil program-text))
                 (original-matrix (when (and protoquil compute-matrix-reps)
-                                   (parsed-program-to-logical-matrix program))))
+                                   (parsed-program-to-logical-matrix program :compress-qubits t))))
```
Comment (appleby) [inline app/src/entry-point.lisp]: The motivation for this change is that when the user requests `---compute-matrix-reps`, we end up comparing `ORIGINAL-MATRIX` to the matrix of the compiled program, and when calling `PARSED-PROGRAM-TO-LOGICAL-MATRIX` on the compiled program, below, we pass `:COMPRESS-QUBITS T`.  If we don't also compress qubits here, there is no guarantee the resulting matrices will be the same size. There are several ways to resolve this:

1. Compress qubits here (current diff represents this choice)
2. Don't compress qubits for `PROCESSED-PROGRAM-MATRIX`, below
3. Call `MATRIX-RESCALE` on the two matrices to ensure they are the same size before passing them to `SCALE-OUT-MATRIX-PHASES` and/or `MATRIX-EQULS-DWIM`.

The downside of (1) is that `ORIGINAL-MATRIX` is meant to be the "known good" base case, so by passing it through `COMPRESS-QUBITS` there is more chance for bugs to invalidate the test. For example, if `COMPRESS-QUBITS` causes `P-P-T-L-M` to generate the same incorrect matrix for both `ORIGINAL-MATRIX` and `PROCESSED-PROGRAM-MATRIX`

The downside to (2) is that it's one less test of `COMPRESS-QUBITS`. For example, if the existing code hadn't been calling `COMPRESS-QUBITS` here, I would have never noticed the bug.

The downside to (3) is I haven't tested it yet and don't know if it actually works :))

File: tests/compiler-hook-tests.lisp [hunk 2]
```diff
@@ -118,9 +120,11 @@ some test programs."
         (format *debug-io* "      Testing file ~a:" (pathname-name file))
         (dolist (architecture (list ':cz ':iswap ':cphase ':piswap ':cnot))
           (format *debug-io* " ~a" architecture)
-          (let ((stats (compare-compiled file architecture)))
-            (when print-stats
-              (format *debug-io* "~a" stats))))
+          (dolist (rewiring-type '(nil :naive :partial :greedy))
```
Comment (appleby) [inline tests/compiler-hook-tests.lisp]: This makes the already-slow `COMPILER-HOOK-TEST` waaaay slower, which is why my intention is to remove this before merging and introduce some smaller, more focused `COMPRESS-QUBITS` tests.

**Comment by notmgsk** [comment]
Without actually reviewing: you have my thanks for introducing some first-class representation of rewiring comments. I think the stuff in `process-program` could be cleaned up with this (and likely other places that scan for rewiring comments).

File: tests/compiler-hook-tests.lisp (hunk 2 above)
Comment (appleby) [reply tests/compiler-hook-tests.lisp]: Thumbs about making COMPILER-HOOK tests slower, or thumbs down about removing this before merging?

File: src/ast.lisp [hunk 3]
```diff
@@ -158,6 +158,89 @@ The keys are typically INSTRUCTION instances and associated values are STRINGs."
...
+  "Parse an exiting REWIRING from REWIRING-STRING."
+  (%parse-single-rewiring rewiring-string +exiting-rewiring-prefix+))
+
+(defun parse-entering/exiting-rewiring (rewiring-string)
+  "Parse entering and exiting REWIRINGs from REWIRING-STRING.
+
+Return (VALUES ENTERING-REWIRING EXITING-REWIRING)."
+  ;; TODO:(appleby) Stolen from PARSED-PROGRAM-TO-LOGICAL-MATRIX. Replace this with something akin to MAKE-REWIRING-FROM-STRING.
+  (let ((*read-eval* nil))
+    (destructuring-bind (entering-vector . exiting-vector)
```
Comment (notmgsk) [inline src/ast.lisp]: I could probably run-it-and-see, but I'm tired: how does this handle a malformed entering/exiting form? e.g. is the error raised by `(parse-entering/exiting-rewiring "Entering/exiting rewiring: ((0 1 2 3) . ())")` intelligible to an end-user?

File: tests/compiler-hook-tests.lisp (hunk 2 above)
Comment (notmgsk) [reply tests/compiler-hook-tests.lisp]: Maybe both! But mostly the slower tests.

File: src/ast.lisp (hunk 3 above)
Comment (appleby) [reply src/ast.lisp]: Currently will blow up more-or-less inscrutably. One of my TODOs above is to rewrite this function to not use `read-from-string` and instead parse it explicitly, a-la `make-rewiring-from-string`.  In that function, I will include a decent error message.

Actually, one question I had (maybe for @ecpeterson) is whether the length of the entering and exiting rewirings always has to match, or would it be possible to have something like

```
(#(0 1 2 3) . #(1 0))
```

Comment (appleby) [reply src/ast.lisp]: ^ For context, I mention this because I noticed there is a `trim-rewiring` function that selects a minimum prefix of rewiring, discarding any tail that is an identity mapping.

Comment (notmgsk) [reply src/ast.lisp]: Ah, it would make sense that redundant info can be elided.

File: src/addresser/rewiring.lisp [hunk 4]
```diff
@@ -271,3 +271,41 @@ BODY as an implicit PROGN."
...
+           :END-ANCHOR)
+         str))
+    (assert matchedp
+            nil
+            "Malformed rewiring pair string: ~@
+             input ~A is not of the form (#(...) . #(...))."
+            str)
+    (let ((first-rewiring-string (aref matches 0))
+          (second-rewiring-string (aref matches 1)))
+      (assert (= (length (aref matches 0)) (length (aref matches 1)))
```
Comment (appleby) [inline src/addresser/rewiring.lisp]: Unclear if the assertion is actually valid and/or required. I opted to err on the side of caution and require the entering/exiting rewirings to be the same length.

File: tests/compiler-hook-tests.lisp (hunk 2 above)
Comment (appleby) [reply tests/compiler-hook-tests.lisp]: These extra compiler hook tests have been removed now. More thorough tests is still a TODO, but I've at least added a new `app/tests/faithfulness-test-files/CNOT-partial-rewiring.quil`.

**Comment by appleby** [comment]
Marking this as "Do Not Review" for now.

The changes in this PR appear to fix the original bug (namely running the "faithfulness" tests in `quilc-tests` with something other than `:naive` as the initial rewiring type). I've added a new test case to the `faithfulness-test-files` that demonstrates this.

However, in the process of writing additional tests, I came across some additional corner cases that need more investigation.

File: src/addresser/rewiring.lisp (hunk 4 above)
Comment (appleby) [reply src/addresser/rewiring.lisp]: Also note that if the entering and exiting rewiring must be the same length, then we could possibly do away with the above regex and speed things up (at the expense of being less forgiving in what we parse) by just splitting the string in half and peeling off outer parens and passing each half to `make-rewiring-from-string`.

File: app/src/entry-point.lisp (hunk 1 above)
Comment (appleby) [reply app/src/entry-point.lisp]: Everything I said above is wrong wrong wrong. Turns out the person who wrote this originally knew what they were doing and I was confused (go figure).

Basically, I was expecting that passing `:compress-qubits t` to `p-p-t-l-m` would result in a matrix of the same size regardless of whether the `parsed-program` you passed has been compiled or not. Wrong!

Long story short: `p-p-t-l-m` has to take the rewirings into account, which means you wind up with a matrix whose size depends on `(max largest-compressed-qubit-index (length (trim-rewiring rewiring))`.  If you run the program through `compress-qubits` first, then the trimmed rewiring might be as long as the highest qubit index in your original source program.

Uncompiled parsed-program's don't have rewirings attached, so `compress-qubits` will produce a matrix of the minimal size based on how many distinct qubit indices appeared in your original parsed-program.

tl;dr

This was correct as it was before: don't `compress-qubits` for the `original-matrix`, but do `compress-qubits` for the `processed-program-matrix`.

**Comment by appleby** [comment]
> Marking this as "Do Not Review" for now.
> 
> The changes in this PR appear to fix the original bug (namely running the "faithfulness" tests in `quilc-tests` with something other than `:naive` as the initial rewiring type). I've added a new test case to the `faithfulness-test-files` that demonstrates this.
> 
> However, in the process of writing additional tests, I came across some additional corner cases that need more investigation.

These "additional corner cases" turned out to be a bug my understanding of how `:compress-qubits t` interacts with `parsed-program-to-logical-matrix` when operating on a `parsed-program` with rewirings attached. See my comment [here](https://github.com/rigetti/quilc/pull/375#discussion_r310744650) for more.

I have one more refactoring TODO to tackle, then I will mark this as ready for review.

File: src/addresser/rewiring.lisp (hunk 4 above)
Comment (appleby) [reply src/addresser/rewiring.lisp]: I just noticed this calculation in `entry-point.lisp` which seems to indicate that entering and exiting rewirings are at least expected to be the same length.

```lisp
(setf (gethash "final_rewiring" statistics)
      (read-from-string
       (subseq (comment instr) length (- (length comment)
                                         (/ (- (length comment) length) 2)
                                         2))))
```

If folks prefer, I could do away with the regex here and do some simpler/faster "parsing by string offset" here as well.

File: app/src/entry-point.lisp [hunk 5]
```diff
@@ -523,58 +523,11 @@ Returns a values tuple (PROCESSED-PROGRAM, STATISTICS), where PROCESSED-PROGRAM
...
-                              (setf (comment (aref (parsed-program-executable-code processed-program)
-                                                   (1- j)))
-                                    (comment instr))))
-                     :unless (typep instr 'halt)
-                       :collect instr)
-               'vector))
-
-        (setf (gethash "topological_swaps" statistics)
-              topological-swaps)
+              (%strip-halts-respecting-rewirings processed-program))
```
Comment (appleby) [inline app/src/entry-point.lisp]: N.B. we now do potentially two passes over `parsed-program-executable-code` to extract the `final_rewiring` and twiddle the `halt`s, whereas we used to do it in a single pass.

However, `extract-final-exit-rewiring-vector` is smart enough to loop over the instructions in reverse, and I assume that the common case is that `processed-program` terminates in with `halt` that has an exit rewiring attached.

File: app/src/entry-point.lisp [hunk 6]
```diff
@@ -638,3 +591,38 @@ Returns a values tuple (PROCESSED-PROGRAM, STATISTICS), where PROCESSED-PROGRAM
                          (quil::lscheduler-calculate-depth lschedule)))))
 
       (values processed-program statistics))))
+
+(defun %strip-halts-respecting-rewirings (processed-program)
```
Comment (appleby) [inline app/src/entry-point.lisp]: Please review the changes to this function with extra care. Part of the reason I split it out into a separate function was so that I could write tests to verify my refactoring didn't break things, but the tests are not exhaustive.

I recommend looking at the diff for https://github.com/rigetti/quilc/pull/375/commits/c57e884584b6426668ccb074f21fdc785dc29de3 to get an easier-to-eyeball apples-to-oranges comparison of the before and after.

Note in particular that I am now checking to make sure rewirings attached to a HALT and it's antecedent are "compatible", and will trigger an assertion if not.

File: app/tests/misc-tests.lisp [hunk 7]
```diff
@@ -43,3 +43,121 @@
...
+             (let ((halt-instr (quil::nth-instr index pp)))
+               (assert (quil::haltp halt-instr))
+               (setf (quil:comment halt-instr)
+                     (quil::make-rewiring-comment :exiting exiting-vector)))
+
+             (when (and (plusp index) (not (null entering-vector)))
+               (setf (quil:comment (quil::nth-instr (1- index) pp))
+                     (quil::make-rewiring-comment :entering entering-vector))))))
+
+    ;; Multiple HALTs in a row are not handled correctly (but should "never" happen).
```
Comment (appleby) [inline app/tests/misc-tests.lisp]: I am basing this assumption on the fact the the original `halt` stripping code did not handle this case.

File: app/src/entry-point.lisp [hunk 8]
```diff
@@ -638,3 +591,38 @@ Returns a values tuple (PROCESSED-PROGRAM, STATISTICS), where PROCESSED-PROGRAM
...
+      :for instr :across (parsed-program-executable-code processed-program)
+      :for prev-instr := (and (plusp j) (quil::nth-instr (1- j) processed-program))
+      :when (and (plusp j) (typep instr 'halt) (comment instr)) :do
+        (multiple-value-bind (this-entering this-exiting) (quil::instruction-rewirings instr)
+          (multiple-value-bind (prev-entering prev-exiting) (quil::instruction-rewirings prev-instr)
+            (assert-rewirings-compatible ':ENTERING instr prev-instr this-entering prev-entering)
+            (assert-rewirings-compatible ':EXITING instr prev-instr this-exiting prev-exiting)
+            ;; Because of the (COMMENT INSTR) in the above :WHEN loop clause, we know that at least
+            ;; one of THIS-ENTERING and THIS-EXITING is non-NIL. Furthermore, due to the
+            ;; ASSERT-REWIRINGS-COMPATIBLE calls we know that either at most one the this/prev
```
Comment (appleby) [inline app/src/entry-point.lisp]:
```suggestion
            ;; ASSERT-REWIRINGS-COMPATIBLE calls we know that either at least one the this/prev
```

File: src/ast.lisp [hunk 9]
```diff
@@ -1313,6 +1431,80 @@ For example,
...
+  (check-type parsed-program parsed-program)
+  (let* ((code (parsed-program-executable-code parsed-program))
+         (length (length code)))
+    (%bounds-check-bidirectional-index index length)
+    (funcall continuation code (%normalize-bidirectional-index index length))))
+
+(defun nth-instr (index parsed-program)
+  "Return the INSTRUCTION at position INDEX in PARSED-PROGRAM's EXECUTABLE-PROGRAM vector.
+
+INDEX can be a \"bidirectional\" index. That is any integer value in the range [-length, length), where length is (LENGTH (PARSED-PROGRAM-EXECUTABLE-CODE PARSED-PROGRAM)).
```
Comment (appleby) [inline src/ast.lisp]: Another, perhaps "lispier", choice is to only allow non-negative INDEX, but accept a `:from-end` keyword option.

File: app/src/entry-point.lisp (hunk 5 above)
Comment (ecpeterson) [reply app/src/entry-point.lisp]: indeed, the only case: within the `protoquil` guard, there can't be any JUMPs, LABELs, ... .

File: src/addresser/rewiring.lisp [hunk 10]
```diff
@@ -271,3 +271,41 @@ BODY as an implicit PROGN."
                      (error "Malformed rewiring string: unexpected token ~a" token))))
                 tokens)))
     (make-rewiring-from-l2p integer-vec)))
+
+(defun make-rewiring-pair-from-string (str)
```
Comment (ecpeterson) [inline src/addresser/rewiring.lisp]: <3

File: src/ast.lisp [hunk 11]
```diff
@@ -158,6 +158,120 @@ The keys are typically INSTRUCTION instances and associated values are STRINGs."
...
+If REWIRING-STRING does not have a valid rewiring comment prefix, signal an error."
+  (cond ((comment-entering-rewiring-p rewiring-string)
+         ':ENTERING)
+        ((comment-exiting-rewiring-p rewiring-string)
+         ':EXITING)
+        ((comment-entering/exiting-rewiring-p rewiring-string)
+         ':ENTERING/EXITING)
+        (t (error "Invalid rewiring comment: ~S" rewiring-string))))
+
+(defun make-rewiring-comment (&key entering exiting)
```
Comment (ecpeterson) [inline src/ast.lisp]: Alternatively (to much of this file), `rewiring-comment` could actually be a struct that stores the tag and the two vectors directly, and then you teach the printer how to turn it into the appropriate string. I don't think we often make the assumption that comments are literally strings—just that they're printable objects.

File: src/ast.lisp [hunk 12]
```diff
@@ -350,6 +464,10 @@ numbers. It must start with a string.")
   #+#:appleby-sufficiently-classy
```
Comment (ecpeterson) [inline src/ast.lisp]: :)

File: src/ast.lisp (hunk 9 above)
Comment (ecpeterson) [reply src/ast.lisp]: This alternative would be my preference, I think, but I also wouldn't stand on it.

File: src/cfg.lisp [hunk 13]
```diff
@@ -525,16 +525,15 @@ Return the following values:
...
-               (format nil "Entering/exiting rewiring: (~a . ~a)"
-                       (rewiring-l2p (basic-block-in-rewiring blk))
-                       (rewiring-l2p (basic-block-out-rewiring blk)))))
+               (make-rewiring-comment :entering (basic-block-in-rewiring blk)
+                                      :exiting (basic-block-out-rewiring blk))))
         (t
          (when (basic-block-in-rewiring blk)
            (setf (comment (aref code-section 0))
-                 (format nil "Entering rewiring: ~a" (rewiring-l2p (basic-block-in-rewiring blk)))))
+                 (make-rewiring-comment :entering (basic-block-in-rewiring blk))))
```
Comment (ecpeterson) [inline src/cfg.lisp]: 👍

File: app/src/entry-point.lisp (hunk 5 above)
Comment (ecpeterson) [reply app/src/entry-point.lisp]: I suspect this assumption would greatly simplify the private method below.

Comment (appleby) [reply app/src/entry-point.lisp]: Yes, that would absolutely simplify things if we can be sure there is at most one HALT stuck on the end.  I realize now that maybe the original code to strip the HALT was just piggy-backing on the pre-existing loop to extract the "final_rewiring", but I assumed it was because there might be multiple HALTs.

File: src/ast.lisp (hunk 9 above)
Comment (appleby) [reply src/ast.lisp]: I think I agree. I will make this change. No sense trying to turn lisp into Python 🐍

File: src/ast.lisp (hunk 11 above)
Comment (appleby) [reply src/ast.lisp]: This is a good idea but unless you feel it's a blocker, I'll add it to the backlog, mostly just so I can move on to Actual Sprint Commitments (tm).

I did consider using some first-class datatype for comments, but went with a minimal gloss over the existing strings/vectors instead.

File: src/ast.lisp (hunk 9 above)
Comment (appleby) [reply src/ast.lisp]: This is done. Changed it take a `:from-end` keyword in the latest commit.

File: src/ast.lisp (hunk 12 above)
Comment (appleby) [reply src/ast.lisp]: 🎩 
🍎

File: app/src/entry-point.lisp (hunk 5 above)
Comment (appleby) [reply app/src/entry-point.lisp]: I take it back. I implemented this change, and it actually made the `%strip-halts...` function slightly longer, not shorter, due to the fact that many of the edge cases fall out of the loop conditions naturally, whereas they have to be explicitly tested if not looping (is the code empty?, does it have a single instruction?, is the final instruction a HALT?, etc).

It's also not really any faster to avoid the loop, since `parsed-program-executable-code` is a `simple-vector`, so you wind up having to copy instructions anyway just to peel off the last one.  What this refactoring *would* do is make the associated tests much simpler to assume it only has to test cases with a single HALT at the end...

But I'm inclined to leave it as is, since it the refactoring doesn't make the code much shorter/faster, and the current version handles multiple HALTs, in case that is ever needed.

File: src/ast.lisp (hunk 11 above)
Comment (appleby) [reply src/ast.lisp]: Filed #381

Comment (ecpeterson) [reply src/ast.lisp]: I don't think this is a blocker. `'pragma` was around for ~2 years before someone thought to subclass it.

File: src/ast.lisp [hunk 14]
```diff
@@ -158,6 +158,120 @@ The keys are typically INSTRUCTION instances and associated values are STRINGs."
   (check-type comment-string string)
   (setf (gethash x **comments**) comment-string))
 
+(a:define-constant +entering-rewiring-prefix+
```
Comment (stylewarning) [inline src/ast.lisp]: I like this refactor, but this in/out rewiring stuff is absurdly complicated for something seemingly so simple.

Comment (appleby) [reply src/ast.lisp]: I think @ecpeterson's idea [here](https://github.com/rigetti/quilc/pull/375#discussion_r312027757) to make rewiring comments a struct with a `print-object` method, rather than always converting to/from strings and vectors would simplify things a fair bit, which is now #381.

Comment (stylewarning) [reply src/ast.lisp]: Well still, the fact that comment data affects semantics in various ways is troubling.
